contracts/chain-adapters/Arbitrum_Adapter.sol

```diff
@@ -233,54 +321,93 @@ contract Arbitrum_Adapter is AdapterInterface, CircleCCTPAdapter {
         }
         // If not, we can use the Arbitrum gateway
         else {
-            uint256 requiredL1CallValue = _contractHasSufficientEthBalance(RELAY_TOKENS_L2_GAS_LIMIT);
-
-            // Approve the gateway, not the router, to spend the hub pool's balance. The gateway, which is different
-            // per L1 token, will temporarily escrow the tokens to be bridged and pull them from this contract.
             address erc20Gateway = L1_ERC20_GATEWAY_ROUTER.getGateway(l1Token);
-            IERC20(l1Token).safeIncreaseAllowance(erc20Gateway, amount);
-
-            // `outboundTransfer` expects that the caller includes a bytes message as the last param that includes the
-            // maxSubmissionCost to use when creating an L2 retryable ticket: https://github.com/OffchainLabs/arbitrum/blob/e98d14873dd77513b569771f47b5e05b72402c5e/packages/arb-bridge-peripherals/contracts/tokenbridge/ethereum/gateway/L1GatewayRouter.sol#L232
-            bytes memory data = abi.encode(L2_MAX_SUBMISSION_COST, "");
-
-            // Note: Legacy routers don't have the outboundTransferCustomRefund method, so default to using
-            // outboundTransfer(). Legacy routers are used for the following tokens that are currently enabled:
-            // - DAI: the implementation of `outboundTransfer` at the current DAI custom gateway
-            //        (https://etherscan.io/address/0xD3B5b60020504bc3489D6949d545893982BA3011#writeContract) sets the
-            //        sender as the refund address so the aliased HubPool should receive excess funds. Implementation here:
-            //        https://github.com/makerdao/arbitrum-dai-bridge/blob/11a80385e2622968069c34d401b3d54a59060e87/contracts/l1/L1DaiGateway.sol#L109
-            if (l1Token == L1_DAI) {
-                // This means that the excess ETH to pay for the L2 transaction will be sent to the aliased
-                // contract address on L2, which we'd have to retrieve via a custom adapter, the Arbitrum_RescueAdapter.
-                // To do so, in a single transaction: 1) setCrossChainContracts to Arbitrum_RescueAdapter, 2) relayMessage
-                // with function data = abi.encode(amountToRescue), 3) setCrossChainContracts back to this adapter.
-                L1_ERC20_GATEWAY_ROUTER.outboundTransfer{ value: requiredL1CallValue }(
-                    l1Token,
-                    to,
-                    amount,
-                    RELAY_TOKENS_L2_GAS_LIMIT,
-                    L2_GAS_PRICE,
-                    data
-                );
+
+            // If custom gas token, call special functions that handle paying with custom gas tokens.
+            if (address(CUSTOM_GAS_TOKEN) != address(0)) {
```

Comment:
Instead of these huge if else blocks, do you think it's worthwhile to just make a different adapter for these rather than squeeze them into the existing arb adapter?

Reply:
hmm there is a lot of code we can re-use though

Reply:
So I guess there are three options:
1. Put them all in the same adapter, as we have here.
2. Copy the code that is identical and make a separate custom gas adapter.
3. Create a common base class to share the common code.

Reply:
I think 2 is the approach that makes the adapters the easiest to reason about, but not super strongly opinionated.

Reply:
2 is also slightly cheaper gas cost wise, i'm ok with that. Makes the diff easier to reason about

Reply:
https://github.com/across-protocol/contracts/pull/585/commits/d6f038d8bf878065cfb9af2821f5428af77cca20